sdk/core/core-lro/src/lroEngine/operation.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AbortSignalLike } from "@azure/abort-controller";
+import { PollOperation, PollOperationState } from "../pollOperation";
+import {
+  PollerConfig,
+  ResumablePollOperationState,
+  LongRunningOperation,
+  GetLroStatusFromResponse,
+  LroResourceLocationConfig,
+  LroStatus
+} from "./models";
+import { getPollingUrl } from "./requestUtils";
+import { createGetLroStatusFromResponse, createInitializeState, createPoll } from "./stateMachine";
+
+export class GenericPollOperation<TResult, TState extends PollOperationState<TResult>>
+  implements PollOperation<TState, TResult> {
+  private poll?: (
+    pollingURL: string,
+    pollerConfig: PollerConfig,
+    getLroStatusFromResponse: GetLroStatusFromResponse<TResult>
+  ) => Promise<LroStatus<TResult>>;
+  private pollerConfig?: PollerConfig;
+  private getLroStatusFromResponse?: GetLroStatusFromResponse<TResult>;
+
+  constructor(
+    public state: TState & ResumablePollOperationState<TResult>,
+    private lro: LongRunningOperation<TResult>,
+    private lroResourceLocationConfig?: LroResourceLocationConfig
+  ) {}
+
+  public setPollerConfig(pollerConfig: PollerConfig): void {
+    this.pollerConfig = pollerConfig;
+  }
+
+  /**
+   * General update function for LROPoller, the general process is as follows
+   * 1. Check initial operation result to determine the strategy to use
+   *  - Strategies: Location, Azure-AsyncOperation, Original Uri
+   * 2. Check if the operation result has a terminal state
+   *  - Terminal state will be determined by each strategy
+   *  2.1 If it is terminal state Check if a final GET request is required, if so
+   *      send final GET request and return result from operation. If no final GET
+   *      is required, just return the result from operation.
+   *      - Determining what to call for final request is responsibility of each strategy
+   *  2.2 If it is not terminal state, call the polling operation and go to step 1
+   *      - Determining what to call for polling is responsibility of each strategy
+   *      - Strategies will always use the latest URI for polling if provided otherwise
+   *        the last known one
+   */
+  async update(options?: {
+    abortSignal?: AbortSignalLike | undefined;
+    fireProgress?: ((state: TState) => void) | undefined;
+  }): Promise<PollOperation<TState, TResult>> {
+    const state = this.state;
+    if (!state.isStarted) {
+      const initializeState = createInitializeState(
+        state,
+        this.lro.requestPath,
+        this.lro.requestMethod
+      );
+      await this.lro.sendInitialRequest(initializeState);
+    }
+
+    if (!state.isCompleted) {
+      if (this.poll === undefined || this.getLroStatusFromResponse === undefined) {
```

Comment:
style nit: it seems like most of these explicit undefined checks are unnecessary, since there are no valid falsy values. 

I find this easier to read:

```suggestion
      if (!this.poll || !this.getLroStatusFromResponse) {
```

Reply:
Done.